IPython/core/magic.py

```diff
@@ -629,10 +632,18 @@ def parse_options(self, arg_str, opt_str, *long_opts, **kw):
             try:
                 opts,args = getopt(argv, opt_str, long_opts)
             except GetoptError as e:
-                raise UsageError('%s ( allowed: "%s" %s)' % (e.msg,opt_str,
-                                        " ".join(long_opts))) from e
-            for o,a in opts:
-                if o.startswith('--'):
+                raise UsageError(
+                    '%s ( allowed: "%s" %s)' % (e.msg, opt_str, " ".join(long_opts))
+                ) from e
+            for o, a in opts:
+                if mode is "string" and preserve_non_opts:
```

Comment:
`mode is "string"` emit a syntax warning, will need to be change to `mode == "string"` in a seubsequent PR

Reply:
see #12920.